xcdat/regridder/accessor.py

```diff
@@ -81,38 +81,67 @@ def grid(self) -> xr.Dataset:
 
         >>> grid = ds.regridder.grid
         """
-        with xr.set_options(keep_attrs=True):
-            coords = {}
-            axis_names: List[CFAxisKey] = ["X", "Y", "Z"]
-
-            for axis in axis_names:
-                try:
-                    data, bnds = self._get_axis_data(axis)
-                except KeyError:
-                    continue
+        axis_names: List[CFAxisKey] = ["X", "Y", "Z"]
 
-                coords[data.name] = data.copy()
+        axis_coords: Dict[str, xr.DataArray] = {}
+        axis_bounds: Dict[str, xr.DataArray] = {}
+        axis_has_bounds: Dict[CFAxisKey, bool] = {}
 
-                if bnds is not None:
-                    coords[bnds.name] = bnds.copy()
-
-        ds = xr.Dataset(coords, attrs=self._ds.attrs)
+        with xr.set_options(keep_attrs=True):
+            for axis in axis_names:
+                coord, bounds = self._get_axis_coord_and_bounds(axis)
+
+                if coord is not None:
+                    axis_coords[str(coord.name)] = coord
+
+                    if bounds is not None:
+                        axis_bounds[str(bounds.name)] = bounds
+                        axis_has_bounds[axis] = True
+                    else:
+                        axis_has_bounds[axis] = False
+
+        # Create a new dataset with coordinates and bounds
+        ds = xr.Dataset(
+            coords=axis_coords,
+            data_vars=axis_bounds,
+            attrs=self._ds.attrs,
+        )
 
-        ds = ds.bounds.add_missing_bounds(axes=["X", "Y", "Z"])
+        # Add bounds only for axes that do not already have them. This
+        # prevents multiple sets of bounds being added for the same axis.
+        # For example, curvilinear grids can have multiple coordinates for the
+        # same axis (e.g., (nlat, lat) for X and (nlon, lon) for Y). We only
+        # need lat_bnds and lon_bnds for the X and Y axes, respectively, and not
+        # nlat_bnds and nlon_bnds.
+        for axis, has_bounds in axis_has_bounds.items():
+            if not has_bounds:
+                ds = ds.bounds.add_bounds(axis=axis)
 
         return ds
 
-    def _get_axis_data(
-        self, name: CFAxisKey
-    ) -> Tuple[xr.DataArray | xr.Dataset, xr.DataArray]:
-        coord_var = get_dim_coords(self._ds, name)
-
-        _validate_grid_has_single_axis_dim(name, coord_var)
-
+    def _get_axis_coord_and_bounds(
+        self, axis: CFAxisKey
+    ) -> Tuple[xr.DataArray | None, xr.DataArray | None]:
         try:
-            bounds_var = self._ds.bounds.get_bounds(name, coord_var.name)
-        except KeyError:
-            bounds_var = None
+            coord_var = get_coords_by_name(self._ds, axis)
+            if coord_var.size == 1:
+                raise ValueError(
+                    f"Coordinate '{coord_var}' is a singleton and cannot be used."
+                )
```

Comment:
Ignore singleton coordinates because we cannot generate bounds for a singleton coordinates yet (e.g., "height" with CF-compliant metadata mapping it to a Z axis).